plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,169 @@
+      The keys are the parser names and the values are either one or several filetypes.
+    '';
 
-        incremental_selection =
-          if cfg.incrementalSelection.enable then
-            {
-              enable = true;
-              keymaps = {
-                init_selection = cfg.incrementalSelection.keymaps.initSelection;
-                node_incremental = cfg.incrementalSelection.keymaps.nodeIncremental;
-                scope_incremental = cfg.incrementalSelection.keymaps.scopeIncremental;
-                node_decremental = cfg.incrementalSelection.keymaps.nodeDecremental;
-              };
-            }
-          else
-            null;
-
-        indent = if cfg.indent then { enable = true; } else null;
-
-        ensure_installed = if cfg.nixGrammars then [ ] else cfg.ensureInstalled;
-        ignore_install = cfg.ignoreInstall;
-        parser_install_dir = cfg.parserInstallDir;
-      } // cfg.moduleConfig;
-    in
-    mkIf cfg.enable {
-      extraConfigLua =
-        (optionalString (cfg.parserInstallDir != null) ''
-          vim.opt.runtimepath:append("${cfg.parserInstallDir}")
-        '')
-        + ''
-          require('nvim-treesitter.configs').setup(${helpers.toLuaObject tsOptions})
-        ''
-        + (optionalString (cfg.languageRegister != { }) ''
-          __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
-
-          for parser_name, ft in pairs(__parserFiletypeMappings) do
-            require('vim.treesitter.language').register(parser_name, ft)
-          end
-        '');
-
-      extraFiles = mkIf cfg.nixvimInjections {
-        "queries/nix/injections.scm" = ''
-          ;; extends
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
-        '';
-      };
-
-      extraPlugins =
-        if cfg.nixGrammars then [ (cfg.package.withPlugins (_: cfg.grammarPackages)) ] else [ cfg.package ];
-      extraPackages = with pkgs; [
-        tree-sitter
-        nodejs
-        cfg.gccPackage
-      ];
-
-      opts = mkIf cfg.folding {
-        foldmethod = "expr";
-        foldexpr = "nvim_treesitter#foldexpr()";
-      };
+    moduleConfig = helpers.defaultNullOpts.mkAttrsOf types.anything { } ''
+      This is the configuration for extra modules. It should not be used directly.
+    '';
+
+    nixGrammars = mkOption {
+      type = types.bool;
+      default = true;
+      description = "Install grammars with Nix";
```

Comment:
You can set `internal = true` and `visible = false` on options that end-users shouldn't use.

You can also set `readOnly = true` if you want to enforce having exactly one definition, though I don't think that applies here.

Reply:
I believe this is a user option to enable / disable the installation of the grammar packages through nix. So a user would use this option if they want nixpkgs managed grammars. This is related to a few TODOs about how we should handle that since the old derivation just changed the defaults when this option was provided but I was wondering if we wanted to add warnings/assertions to those settings if this is enabled.

Reply:
Yes, this option should belong in `extraOptions`.